hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java

```diff
@@ -0,0 +1,190 @@
+ * and executing some operations when the first connection request returns a valid
+ * connection.
+ *
+ * Useful to make sure an application can start even if the database is nto yet ready or present
+ * while still be able to execute schema generation orders.
+ *
+ * @author Emmanuel Bernard emmanuel@hibernate.org
+ */
+public class DelayerConnectionProvider implements
+		ConnectionProvider,
+		Configurable,
+		ServiceRegistryAwareService,
+		Stoppable {
+
+	private static final Logger LOGGER = Logger.getLogger( DelayerConnectionProvider.class );
+
+	private final ConnectionProvider connectionProvider;
+	private ServiceRegistry serviceRegistry;
+	// volatile due to the double checked lock
+	private volatile List<Runnable> operations = new ArrayList<>();
+	private final ReentrantLock lock = new ReentrantLock();
+
+	public DelayerConnectionProvider(ConnectionProvider connectionProvider) {
+		this.connectionProvider = connectionProvider;
+	}
+
+	/**
+	 * Operations can only be added before the {@code SessionFactory} is fully built.
+	 * This is to prevent concurrent additions of operations.
+	 * The operation must use the {@link ServiceRegistry} provided to retrieve
+	 * {@link ConnectionProvider} to get a connection.
+	 *
+	 * The operation will try and be executed eagerly, we might be lucky.
+	 *
+	 * @param operation on a connection ready database
+	 */
+	public void addOperation(Runnable operation) {
+		operations.add( operation );
+		tryAndExecuteOperations();
+	}
+
+	/**
+	 * Try to eagerly flush the operations if a connection is available
+	 */
+	private void tryAndExecuteOperations() {
+		// It would be more elegant to refactor the operation lock from getConnection is a shared method
+		// but I prefer the JVM to have more chances to keep getConnection() inlined
+		try {
+			Connection c = getConnection();
+			closeConnection( c );
+			if (operations == null) {
+				// operation execution has succeeded, but subsequent ones might happen
+				// remember, we are not multi-threaded here as we are bootstrapping Hibernate ORM
+				operations = new ArrayList<>();
+			}
+		}
+		catch (Exception e) {
+			// we don't care if there is a failure
+		}
+	}
+
+	@Override
+	public void injectServices(ServiceRegistryImplementor serviceRegistry) {
+		this.serviceRegistry = serviceRegistry;
+		if ( connectionProvider instanceof ServiceRegistryAwareService ) {
+			( (ServiceRegistryAwareService) connectionProvider ).injectServices( serviceRegistry );
+		}
+	}
+
+	@Override
+	public void configure(Map configurationValues) {
+		if ( connectionProvider instanceof Configurable ) {
+			Configurable configurableConnectionProvider = (Configurable) connectionProvider;
+			configurableConnectionProvider.configure( configurationValues );
+		}
+	}
+
+	@Override
+	public Connection getConnection() throws SQLException {
+		Connection connection;
+		try {
+			connection = connectionProvider.getConnection();
```

Comment:
A more concrete example: suppose this is invoked in the scope of a transaction for thread A. Another thread B just got the connection, and is now going to execute DDL operations. Thread A is waiting for those to complete, but when it gets the green light it might not have the tables yet, as its transaction started before they were created.